python/src/com/jetbrains/python/psi/types/PyImportedModuleType.java

```diff
@@ -40,23 +34,39 @@ public PyImportedModuleType(@NotNull PyImportedModule importedModule) {
-    final PsiElement resolved = myImportedModule.resolve();
-    if (resolved != null) {
-      final PsiFile containingFile = location != null ? location.getContainingFile() : null;
-      List<PsiElement> elements = Collections.singletonList(ResolveImportUtil.resolveChild(resolved, name, containingFile, false, true,
-                                                                                           false));
-      final PyImportElement importElement = myImportedModule.getImportElement();
-      final PyFile resolvedFile = PyUtil.as(resolved, PyFile.class);
-      if (location != null && importElement != null && PyUtil.inSameFile(location, importElement) &&
-          ResolveImportUtil.getPointInImport(location) == PointInImport.NONE && resolved instanceof PsiFileSystemItem &&
-          (resolvedFile == null || !PyUtil.isPackage(resolvedFile) || resolvedFile.getElementNamed(name) == null)) {
-        final List<PsiElement> importedSubmodules = PyModuleType.collectImportedSubmodules((PsiFileSystemItem)resolved, location);
-        if (importedSubmodules != null) {
-          final Set<PsiElement> imported = Sets.newHashSet(importedSubmodules);
-          elements = ContainerUtil.filter(elements, element -> imported.contains(element));
+
+    List<PsiElement> importedModuleCandidates = myImportedModule.multiResolve();
+    for (PsiElement moduleLocation : importedModuleCandidates) {
+      final PsiFileSystemItem resolvedModuleOrPackage = PyUtil.as(moduleLocation, PsiFileSystemItem.class);
+      if (resolvedModuleOrPackage != null) {
+        List<? extends RatedResolveResult> results =
+          PyModuleType.resolveMemberInPackageOrModule(myImportedModule, resolvedModuleOrPackage, name, location, resolveContext);
+        if (!ContainerUtil.isEmpty(results)) {
+          return results;
+        }
+      }
+    }
+
+    for (PsiElement moduleLocation : importedModuleCandidates) {
+      final PsiFileSystemItem resolvedModuleOrPackage = PyUtil.as(moduleLocation, PsiFileSystemItem.class);
+      if (resolvedModuleOrPackage == null) {
+        continue;
+      }
+      //if it's a non-package module or a synthetic import element
+      if (!PyUtil.isPackage(resolvedModuleOrPackage, location) || myImportedModule.getImportElement() == null) {
+        PsiFile resolvingFromFile = Optional.ofNullable(location).map(l -> l.getContainingFile()).orElse(null);
+        List<RatedResolveResult> fallbackResults =
+          ResolveImportUtil.resolveChildren(resolvedModuleOrPackage, name, resolvingFromFile, false, true,
+                                            false, false);
+
+        if (!ContainerUtil.isEmpty(fallbackResults)) {
+          ResolveResultList ret = new ResolveResultList();
+          fallbackResults.stream()
+                         .map(res -> new ImportedResolveResult(res.getElement(), res.getRate(), myImportedModule.getImportElement()))
+                         .forEach(ret::add);
+          return ret;
```

Comment:
Consider using our `ContainerUtil` classes
E.g. this code could be rewritten as `return ContainerUtil.map(fallbackResults, res -> new ImportedResolveResult(res.getElement(), res.getRate(), myImportedModule.getImportElement()));`